File: apps/guzo-api/src/main/java/et/guzo/web/controller/FamilyController.java
package et.guzo.web.controller;

import et.guzo.common.ApiResponse;
import et.guzo.domain.entity.FamilyMember;
import et.guzo.security.RoleChecker;
import et.guzo.security.SecurityUtil;
import et.guzo.service.FamilyService;
import et.guzo.web.dto.FamilyLinkRequest;
import et.guzo.web.dto.FamilyMemberResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/family")
@RequiredArgsConstructor
public class FamilyController {

    private final FamilyService familyService;

    @GetMapping("/{ownerUserId}")
    public ApiResponse<List<FamilyMemberResponse>> list(@PathVariable String ownerUserId) {
        var user = SecurityUtil.requireUser();
        if (!user.getId().equals(ownerUserId) && !RoleChecker.isAdmin(user)) {
            throw et.guzo.common.ApiException.forbidden("Not allowed");
        }
        List<FamilyMemberResponse> items = familyService.listForOwner(ownerUserId).stream()
            .map(FamilyMemberResponse::from)
            .toList();
        return ApiResponse.ok(items);
    }

    @PostMapping("/link")
    public ApiResponse<FamilyMemberResponse> link(@Valid @RequestBody FamilyLinkRequest body) {
        var user = SecurityUtil.requireUser();
        if (!user.getId().equals(body.ownerUserId()) && !RoleChecker.isAdmin(user)) {
            throw et.guzo.common.ApiException.forbidden("Not allowed");
        }
        FamilyMember fm = familyService.link(body.ownerUserId(), body.memberUserId(), body.relation(), body.label());
        return ApiResponse.ok(FamilyMemberResponse.from(fm), "Family member linked");
    }
}

File: apps/server/src/modules/family/family.controller.ts
import type { Request, Response } from 'express';
import { asyncHandler } from '../../utils/asyncHandler.js';
import { ok, created } from '../../utils/ApiResponse.js';
import { familyService } from './family.service.js';
import { ApiError } from '../../utils/ApiError.js';

export const familyController = {
  list: asyncHandler(async (req: Request, res: Response) => {
    const ownerUserId = req.params.ownerUserId;
    if (req.user!.id !== ownerUserId && !req.user!.roles.some((r) => ['ADMIN', 'SUPER_ADMIN'].includes(r))) {
      throw ApiError.forbidden('Not allowed');
    }
    const items = await familyService.listForOwner(ownerUserId);
    return ok(res, items);
  }),

  link: asyncHandler(async (req: Request, res: Response) => {
    const { ownerUserId, memberUserId, relation, label } = req.body as {
      ownerUserId: string;
      memberUserId: string;
      relation?: string;
      label?: string;
    };
    if (req.user!.id !== ownerUserId && !req.user!.roles.some((r) => ['ADMIN', 'SUPER_ADMIN'].includes(r))) {
      throw ApiError.forbidden('Not allowed');
    }
    const item = await familyService.link(ownerUserId, memberUserId, relation ?? 'OTHER', label);
    return created(res, item, 'Family member linked');
  }),
};
